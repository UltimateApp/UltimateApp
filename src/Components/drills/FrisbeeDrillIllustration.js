import React, { useState, useEffect, useRef } from 'react';
import { View, StyleSheet, Text, Dimensions, TouchableOpacity, SafeAreaView } from 'react-native';
import { WebView } from 'react-native-webview';
import Carousel, { Pagination } from 'react-native-snap-carousel';
import { MaterialCommunityIcons } from '@expo/vector-icons';

import Animation from '../animation/Animation';
import VimeoVideo from '../shared/VimeoVideo';
import { IllustrationType } from '../../Fixtures/config';
import theme from '../../styles/theme.style';
import Drill from '../animation/Drill';

const screenDimension = Dimensions.get('window');

const FrisbeeDrillIllustration = props => {
  const [activeIndex, setActiveIndex] = useState(0);
  const currentStep = props.drill.steps[activeIndex];

  const carouselRef = useRef(null);

  // back to 0 when drill changes
  useEffect(() => {
    setActiveIndex(0);
  }, [props.drill]);

  const renderTitle = (title, index) => {
    const isFirstStep = index === 0;
    const isLastStep = index === props.drill.steps.length - 1;
    return (
      <View style={styles.line}>
        <View>
          {!isFirstStep && (
            <TouchableOpacity
              onPress={() => {
                carouselRef.current.snapToPrev();
              }}
            >
              <MaterialCommunityIcons name="chevron-double-left" color={theme.COLOR_PRIMARY} size={26} />
            </TouchableOpacity>
          )}
        </View>
        <Text style={styles.title}>{title}</Text>
        <View>
          {!isLastStep && (
            <TouchableOpacity
              onPress={() => {
                carouselRef.current.snapToNext();
              }}
            >
              <MaterialCommunityIcons name="chevron-double-right" color={theme.COLOR_PRIMARY} size={26} />
            </TouchableOpacity>
          )}
        </View>
      </View>
    );
  };

  const displayYoutube = ({ illustrationSource }) => {
    return (
      <View style={styles.contentWrapper}>
        <View style={{ height: 250 }}>
          <WebView
            source={{
              uri: illustrationSource,
            }}
          />
        </View>
      </View>
    );
  };

  const displayVimeo = ({ illustrationSource, sounds }) => {
    return (
      <View style={styles.contentWrapper}>
        <View style={[{ height: 250 }, styles.videoAlone]}>
          <VimeoVideo vimeoId={illustrationSource} screenWidth={screenDimension.width} sounds={sounds} />
        </View>
      </View>
    );
  };

  const displayAnimation = ({ illustrationSource, instruction }) => {
    return (
      <>
        <View style={styles.contentWrapper}>
          <Animation widthRatio={1} heightRatio={1 / 2} animation={new Drill(illustrationSource)} />
        </View>
        <Text style={styles.instruction}>{instruction}</Text>
      </>
    );
  };

  const renderStep = ({ item, index }) => {
    return (
      <>
        {renderTitle(item.title, index)}
        <View style={styles.pagination}>{renderPagination()}</View>
        {item.illustrationType === IllustrationType.ANIMATION && displayAnimation(item)}
        {item.illustrationType === IllustrationType.YOUTUBE && displayYoutube(item)}
        {item.illustrationType === IllustrationType.VIMEO && displayVimeo(item)}
      </>
    );
  };

  const renderPagination = () => {
    return (
      <Pagination
        dotsLength={props.drill.steps.length}
        activeDotIndex={activeIndex}
        containerStyle={{
          backgroundColor: theme.BACKGROUND_COLOR_LIGHT,
          paddingVertical: 0,
        }}
        dotStyle={{
          width: 8,
          height: 8,
          borderRadius: 5,
          marginHorizontal: 5,
          backgroundColor: theme.MAIN_COLOR,
        }}
        inactiveDotStyle={{
          backgroundColor: theme.COLOR_SECONDARY,
        }}
        inactiveDotOpacity={0.4}
        inactiveDotScale={0.6}
        tappableDots
        carouselRef={carouselRef}
      />
    );
  };

  return (
    <SafeAreaView style={styles.container}>
      <Carousel
        layout="default"
        ref={carouselRef}
        data={props.drill.steps}
        sliderWidth={screenDimension.width}
        itemWidth={screenDimension.width}
        renderItem={renderStep}
        onSnapToItem={index => setActiveIndex(index)}
      />
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  instruction: {
    marginBottom: 10,
    marginHorizontal: 10,
    fontSize: theme.FONT_SIZE_MEDIUM,
    color: theme.COLOR_PRIMARY,
  },
  videoAlone: {
    flex: 1,
  },
  pagination: {
    paddingVertical: 10,
  },
  line: {
    height: 40,
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginHorizontal: 30,
    alignItems: 'center',
  },
  title: {
    fontSize: theme.FONT_SIZE_MEDIUM,
    color: theme.COLOR_PRIMARY,
    textAlign: 'center',
  },
  contentWrapper: { minHeight: 375 },
  centerContainer: {
    flex: 1,
    alignItems: 'center',
  },
});

export default FrisbeeDrillIllustration;
