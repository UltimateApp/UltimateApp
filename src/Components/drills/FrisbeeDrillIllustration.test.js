import React from 'react';
import renderer from 'react-test-renderer';
import { render, cleanup } from 'react-native-testing-library';
import { createDrill } from '../../Fixtures/TestFixtures';

import FrisbeeDrillIllustration from './FrisbeeDrillIllustration';

jest.mock('react-native/Libraries/Animated/src/NativeAnimatedHelper');

describe('<FrisbeeDrillIllustration />', () => {
  it('renders correctly', () => {
    const drill = createDrill();
    const { toJSON } = render(<FrisbeeDrillIllustration drill={drill} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
