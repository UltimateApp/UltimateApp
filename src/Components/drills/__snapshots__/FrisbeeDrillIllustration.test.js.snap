// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FrisbeeDrillIllustration /> renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    activeAnimationOptions={null}
    activeAnimationType="timing"
    activeSlideAlignment="center"
    activeSlideOffset={20}
    apparitionDelay={0}
    automaticallyAdjustContentInsets={false}
    autoplay={false}
    autoplayDelay={1000}
    autoplayInterval={3000}
    callbackOffsetMargin={5}
    containerCustomStyle={Object {}}
    contentContainerCustomStyle={Object {}}
    contentContainerStyle={
      Array [
        Object {
          "paddingLeft": 0,
          "paddingRight": 0,
        },
        Object {},
      ]
    }
    data={
      Array [
        Object {
          "illustrationSource": "https://www.youtube.com/embed/JkVHrA5o23o",
          "illustrationType": "youtube",
          "subtitle": "This is a subtitle",
          "title": "Warmup",
        },
      ]
    }
    decelerationRate="fast"
    directionalLockEnabled={true}
    disableVirtualization={false}
    enableMomentum={false}
    enableSnap={true}
    firstItem={0}
    getItem={[Function]}
    getItemCount={[Function]}
    hasParallaxImages={false}
    horizontal={true}
    inactiveSlideOpacity={0.7}
    inactiveSlideScale={0.9}
    inactiveSlideShift={0}
    initialNumToRender={6}
    invertStickyHeaders={false}
    inverted={false}
    itemWidth={750}
    keyExtractor={[Function]}
    layout="default"
    lockScrollTimeoutDuration={1000}
    lockScrollWhileSnapping={false}
    loop={false}
    loopClonesPerSide={3}
    maxToRenderPerBatch={13}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onResponderRelease={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    onSnapToItem={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    overScrollMode="never"
    pinchGestureEnabled={false}
    removeClippedSubviews={true}
    renderItem={[Function]}
    scrollEnabled={true}
    scrollEventThrottle={1}
    scrollsToTop={false}
    shouldOptimizeUpdates={true}
    showsHorizontalScrollIndicator={false}
    showsVerticalScrollIndicator={false}
    slideStyle={Object {}}
    sliderWidth={750}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flexDirection": "row",
        "opacity": 0,
        "width": 750,
      }
    }
    swipeThreshold={20}
    updateCellsBatchingPeriod={50}
    useScrollView={false}
    vertical={false}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={13}
  >
    <View>
      <View
        onLayout={[Function]}
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            null,
          ]
        }
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "opacity": 1,
              "transform": Array [
                Object {
                  "scale": 1,
                },
              ],
              "width": 750,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "height": 40,
                "justifyContent": "space-between",
                "marginHorizontal": 30,
              }
            }
          >
            <View />
            <Text
              style={
                Object {
                  "color": "#232b2b",
                  "fontSize": 16,
                  "textAlign": "center",
                }
              }
            >
              Warmup
            </Text>
            <View />
          </View>
          <View
            style={
              Object {
                "paddingVertical": 10,
              }
            }
          />
          <RCTScrollView>
            <View>
              <View
                style={
                  Object {
                    "minHeight": 375,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "height": 250,
                    }
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "flex": 1,
                          "overflow": "hidden",
                        },
                        undefined,
                      ]
                    }
                  >
                    <RNCWebView
                      cacheEnabled={true}
                      javaScriptEnabled={true}
                      messagingEnabled={false}
                      onContentProcessDidTerminate={[Function]}
                      onHttpError={[Function]}
                      onLoadingError={[Function]}
                      onLoadingFinish={[Function]}
                      onLoadingProgress={[Function]}
                      onLoadingStart={[Function]}
                      onMessage={[Function]}
                      onShouldStartLoadWithRequest={[Function]}
                      source={
                        Object {
                          "uri": "https://www.youtube.com/embed/JkVHrA5o23o",
                        }
                      }
                      style={
                        Array [
                          Object {
                            "flex": 1,
                            "overflow": "hidden",
                          },
                          Object {
                            "backgroundColor": "#ffffff",
                          },
                          undefined,
                        ]
                      }
                      useSharedProcessPool={true}
                    />
                  </View>
                </View>
              </View>
            </View>
          </RCTScrollView>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
